canonicalization overhaul part 3:  Types, Templates, and Public/Private

jcollins-g opened the pull request:
Going to split this out and clean up some of the architectural duct tape here, but creating a PR for reference so @devoncarew and @keertip can see where this is going.

Investigating #1512 resulted in the detection of a regression in #1514, related to the generic type support recently added in #1506 and related PRs.  It turns out this is badly broken -- as indicated in #1514 there are some highly visible docs that have confusing types listed since it's normal for different classes type parameters to be `T`.

Fixing this pulled on some threads related once again to canonicalization introduced in #1319, with refinements in #1368 and #1390.  In order to resolve types correctly, dartdoc's ElementType class has to be instantiated properly, with the right ModelElement and type.  To do that, ModelElements have to keep track of the original Member objects that contain type resolution information from the analyzer.  And finally, in order to handle referring to elements defined in non-public libraries but reexported elsewhere, we need to have a better handle on what is public vs. non-public, and canonical vs. not-canonical.

The original pre-canonicalization dartdoc documented everything it saw if it was public, and didn't document it if it wasn't public.  This made sense since dartdoc didn't try to dedupe as requested in issue #1090, so resolved types could always be displayed.  Dartdoc also didn't need to distinguish between Member and Element objects and always used them the same, because issues related to generic types hadn't yet revealed problems like dart-lang/sdk#29600.

However, after canonicalization, there were various hacks that tried to keep type resolution mostly functional, though resulting in some inconsistent behavior noted in #1512.  Still, until we added [support for anonymous functions](https://github.com/dart-lang/dartdoc/pull/1506) & [working generic types](https://github.com/dart-lang/dartdoc/pull/1509), these hacks were good enough.  But the whole mess finally came to a head there and it became impossible to fix #1514 without either breaking the function support or reworking the base data structures to include more information, including information about non-public ModelElements contained in classes, libraries, etc.

Adding non-public items meant fixing the private/public distinction inside ModelElements.  This now results in a new property, isDocumented, and new hasFoo / allFoo equivalents explicitly filtering for public items if needed.  (hasPublicFoo, allPublicFoo) with the default being that all elements are included.  Fixing the templates for that was virtually impossible with the old version of mustache4dart which silently failed, but the new version obeys the "fail on missing" flag so I upgraded that and then promptly discovered a pile of existing template bugs as well, fixing those.

Now that all objects are included so that type resolution works and the isDocumented call checks canonicalization, warnings became more likely to happen on non-documented elements because we traverse them now.   So introduced some infinite-recursion prevention in the warning system and added isDocumented checks for a couple of the non-canonicalization-related warnings typically generated by non-public elements.  This results in a net win, hiding many warnings we had that were nonsensical from before this change (partial fix for #1412) as well as squelching new warnings generated by the type resolution fixes.

The planned breakout of this change will be to submit the pieces in the following order if I can:

- Warnings improvement
- Public/private distinction, mustache4dart upgrade, and template fixes
- ModelElement instantiation and Member tracking
- Type resolution fixes

Disentangling them along these lines will take a bit but should make it easier to review.

kevmoo commented on lib/src/html/html_generator_instance.dart:
```diff
@@ -96,86 +96,78 @@ class HtmlGeneratorInstance implements HtmlOptions {
-        if (!property.isCanonical) continue;
+      for (var property in lib.properties.where((p) => p.isDocumented)) {
+        if (property.name == 'topLevelVariable')
+          1+1;
```
???????

kevmoo replied:
????

kevmoo commented on lib/src/html/html_generator_instance.dart:
```diff
@@ -96,86 +96,78 @@ class HtmlGeneratorInstance implements HtmlOptions {
+        if (property.name == 'topLevelVariable')
+          1+1;
+        if (property.linkedReturnType == '')
+          1+1;
```
Ditto

kevmoo commented on lib/src/markdown_processor.dart:
```diff
@@ -390,6 +390,7 @@ Element _findRefElementInLibrary(String codeRef, ModelElement element,
   assert(element != null);
   assert(element.package.allLibrariesAdded);
 
+
```
??? Dartfmt won't like this...

jcollins-g commented:
To be clear, this is WIP and not intended to land.  Mostly this will be reference material for the breakout pieces as I submit those.

devoncarew commented:
Taking a spin through now just to get familiar -

devoncarew commented on lib/src/html/html_generator_instance.dart:
```diff
@@ -96,86 +96,78 @@ class HtmlGeneratorInstance implements HtmlOptions {
     generatePackage();
 
-    for (var lib in package.libraries) {
+    for (var lib in package.libraries.where((l) => l.isDocumented)) {
```
Consider putting `isDocumented` onto Documentable, and having a utility method that would filter a list of Documentable objects to just the ones that were documented.

devoncarew replied:
(I see a lot of `.where((o) => o.isDocumented` below)

jcollins-g replied:
done

kevmoo commented:
Ran this through the SDK – just curious. Sooo much better. Love seeing all of random toString and hashCode members gone. Also great to see links to Stream<List<int>> in generics. Very nice.

Two things I noticed.

1) Some private members show up

Compare https://api.dartlang.org/be/151323/dart-io/FileSystemEvent-class.html CONSTANTS to
<img width="238" alt="screen shot 2017-10-26 at 1 08 23 pm" src="https://user-images.githubusercontent.com/17034/32074692-1bfe7eec-ba4f-11e7-8115-e01665979f2a.png">

2) Saw a few cases where `Implements` links are dead

Look at https://api.dartlang.org/be/151323/dart-io/ContentType-class.html
See how the `HeaderValue` is live? It's not w/ this PR

I have no clue if you're already aware of these – sorry for the noise if you are.

Otherwise, super excited to see this change get in!

kevmoo commented on lib/src/warnings.dart:
```diff
@@ -256,12 +256,14 @@ class PackageWarningCounter {
+    if (_warningCounts.isEmpty) return 0;
     return _warningCounts.keys
         .map((w) => options.asErrors.contains(w) ? _warningCounts[w] : 0)
         .reduce((a, b) => a + b);
```
using `fold` here would let you avoid the extra case

kevmoo commented on lib/src/warnings.dart:
```diff
@@ -256,12 +256,14 @@ class PackageWarningCounter {
+    if (_warningCounts.isEmpty) return 0;
     return _warningCounts.keys
         .map((w) =>
             options.asWarnings.contains(w) && !options.asErrors.contains(w)
```
Ditto below – replace `reduce` with `fold`...at least I think

jcollins-g commented:
With the exception of flutter's dartdoc script needing to be updated for new parameters and code cleanup, I've finished making the part 3 overhaul function for Flutter, Angular, and the Dart SDK with head and stable versions of analyzer, as well as addressing the issues @kevmoo mentioned and many more.

If it wasn't a breaking change before, it definitely is now.  New flags and new naming of anonymous libraries make this a big change.

In order to fix the complex packages without losing functionality, several more changes were required.

- #1535 and #1529 had to be addressed.  Lots of previously unlinkable things becoming linkable in the overhaul resulted in this bug getting worse, hitting cases where generating invalid HTML in the summary case where the string was truncated.
- #1526 had to be dealt with.  Anonymous libraries with the same base name inside the same package is now a thing in Angular.  Dartdoc now places anonymous libraries on disk in such a way to avoid collisions, but keeps the user-visible names of the libraries as they were.
- Basic package boundary awareness has been added to dartdoc.  Dartdoc now knows a little bit about package boundaries in canonicalization, and --auto-include-dependencies now traverses package rather than library dependencies.  A new --exclude-packages flag allows for the exclusion of packages when using --auto-include-dependencies.   This was necessary to squelch bogus warnings.
- Package boundary awareness means fewer ambiguous doc reference warnings when using --auto-include-dependencies.  dartdoc now knows to stay local to the package when it is ambiguous which symbol a reference links to.
- Package boundary awareness means all useless "no canonical found" warnings have been silenced at last.  There's only one left now in the SDK that seems to be legitimate (a public class AttributeClassSet  inherits from a impl class that's only in dart:html_common).

Code here still needs to be cleaned up, including removing my persistent breakpoints (the "1+1"s), and refactoring package construction so that the tests can use the same code as the mainline dartdoc for recursion.

Additional pieces for the planned split in addition to what I already indicated:
- #1529 and #1535
- Disk layout of anonymous libraries
- Some aspects of package boundary awareness

kevmoo commented:
FWIW, I went through the cases I'd seen. All cleared up. Everything looks so much better. Awesome work!

devoncarew commented:
@jcollins-g, just to confirm:

> The planned breakout of this change will be to submit the pieces in the following order if I can:

Still planned (the-breakup-into-pieces)?

jcollins-g commented:
@devoncarew  Yes.  There will be a couple more pieces now, sadly, but that is still planned.

jcollins-g commented:
Oh, and for those perusing the code, there's been an additional change here that I was trying to avoid but it was just getting too hard to deal with the cut-and-pastiness of Dartdoc.  (Not anyone's fault  -- this is what happens when a small tool grows without enough focused time on updating the design to match).

I've begun selectively factoring out common sections of code and interfaces into mixins and abstract classes, hopefully in a way that makes some sense.  There's now a Canonicalization class for adding Dartdoc canonicalization support, a Privacy class for elements that support isPublic and its friends, a TypeParameters class for elements with type parameters, and so on.

This hasn't yet been addressed systematically because I was really trying to just fix one bug here and I'm only factoring out code that I'm touching.  But hopefully I will have time to follow up and address the larger structural changes Dartdoc needs.

kevmoo commented:
> I've begun selectively factoring out common sections of code and interfaces...

🎉 Yay for maintainability!

jcollins-g commented:
Starting the breakout now, first piece will be some of the separable warnings changes.

jcollins-g commented:
I've managed to break out a few pieces, but disentangling changes is starting to become difficult, as it's getting very hard to identify non-mutually-dependent pieces of any size.  A few things I thought might be separable, like elements of the private/public distinction, can't be extracted without dragging most or all over the overhaul along and still have dartdoc pass tests and generate valid docs.   So, we're going to try reviewing this as it is now.  Open for comments!

devoncarew commented on lib/dartdoc.dart:
```diff
@@ -9,33 +9,20 @@ import 'dart:async';
-import 'package:analyzer/source/sdk_ext.dart';
-import 'package:analyzer/src/context/builder.dart';
-import 'package:analyzer/src/dart/sdk/sdk.dart';
+import 'package:analyzer/analyzer.dart';
```
That's convenient :)

devoncarew commented on lib/dartdoc.dart:
```diff
@@ -78,108 +65,98 @@ Future<List<Generator>> initGenerators(String url, String relCanonicalPrefix,
+          .where((_Error error) => error.isError)
+          .toList()
+            ..sort();
+      // TODO(jcollins-g): Why does the SDK have analysis errors?  Annotations
```
Perhaps also track as an issue?

jcollins-g replied:
done, and issue added to relevant TODOs.

devoncarew commented on lib/src/element_type.dart:
```diff
@@ -58,7 +71,10 @@ class ElementType {
 
-  String get name => _type.name;
+  String get name {
+    if (_type.name == null) return _type.element.name;
```
or, `return _type.name ?? _type.element.name;`?

devoncarew commented on lib/src/model.dart:
```diff
@@ -221,8 +259,9 @@ class Accessor extends ModelElement
-  Accessor(PropertyAccessorElement element, Library library)
-      : super(element, library);
+  Accessor(
+      PropertyAccessorElement element, Library library, Member originalMember)
```
I see some subclasses pass `null` in for `originalMember`; perhaps make it an `[optional]` param?

devoncarew commented on lib/src/model.dart:
ditto re: possibly making `_originalMember` optional

devoncarew commented on lib/src/model.dart:
```diff
@@ -1678,7 +1809,61 @@ class Library extends ModelElement {
+  Iterable<TopLevelVariable> get publicConstants => filterNonPublic(constants);
+
+  String _dirName;
+  String get dirName {
```
Thinking about a pattern for this; not sure this helps, but here's one idea:

```dart
String get dirName {
  return _dirName ?? (_dirName = () {
    return ... ; // init code...
  }());
}
```

devoncarew commented on lib/src/model.dart:
```diff
@@ -4548,3 +4864,313 @@ class TypeParameter extends ModelElement {
 }
+
+/// Everything you need to instantiate a Package object for documenting.
+class PackageBuilder {
```
+1, this, class `Canonicalization`, class `Privacy` - I think the modeling improvements here will really help.

devoncarew commented on lib/src/model.dart:
```diff
@@ -4548,3 +4864,313 @@ class TypeParameter extends ModelElement {
+  AnalysisContext get context {
+    if (_context == null) {
+      // TODO(jcollins-g): fix this so it actually obeys analyzer options files.
+      var options = new AnalysisOptionsImpl()..enableAssertInitializer = true;
```
@bwilkerson - Janice has changed dartdoc so that it only creates an analysis context in one place now. We should look into unifying with how the analysis server creates contexts - we have several outstanding issues as dartdoc doesn't pick up the analysis options settings from projects.

devoncarew commented on test/src/utils.dart:
```diff
@@ -39,53 +32,43 @@ void delete(Directory dir) {
+          [],
+          [],
+          true,
+          withAutoIncludedDependencies)
```
a trailing comma here (and some other places with less than ideal formatting) may make dartfmt format this better

devoncarew replied:
(I find that trailing commas can make things this split lines dartfmt better)

jcollins-g commented:
Disabling dartdoc's flutter travis for now, since they need to pass different parameters to dartdoc for the new version to work.  I'm checking Flutter manually for now.

kevmoo commented:
Well done, @jcollins-g !!!
